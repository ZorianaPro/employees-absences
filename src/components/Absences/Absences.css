.Absences .Absences-Container {
  position: relative;
  height: 100%;
  width: 100%;
  padding: 2.5em 0 9em;
  display: flex;
  margin: 2em 0;
}

.Absences .react-calendar {
  width: 30em;
  max-height: 20em;
}

.Absence {
  display: flex;
  flex-direction: row;
  align-items: center;
  margin: 0 2em 1em 2em;
  border: 1px solid var(--colorRaisinBlack);
  border-radius: 0.5em;
  cursor: pointer;
}

.Absence .Absence-Info > p{
  padding: 0 0.5em;
}

li {
  list-style: none;
  padding: 0;
  display: block;
  margin: 0;
}