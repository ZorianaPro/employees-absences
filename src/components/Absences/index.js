import React, { useEffect, useState, useCallback } from 'react';
import { Link, Redirect, useLocation } from 'react-router-dom';
import iCalendarService from '../../services/iCalendar';
import absencesService from '../../services/absences';
import membersService from '../../services/members';
import Calendar from 'react-calendar';
import SVGSpriteSheet from '../SVGSpriteSheet';
import Overlay from '../Overlay';
import Member from '../Member';
import AbsenceInfo from '../AbsenceInfo';
import 'react-calendar/dist/Calendar.css';
import './Absences.css';

const Absences = () => {
  const [absences, setAbsences] = useState([]);
  const [absencesToShow, setAbsencesToShow] = useState([]);
  const [selectedView, setSelectedView] = useState('day');
  const [activeStartDate, setActiveStartDate] = useState(new Date(2017, 1, 1, 0, 0));
  const [shouldOpenOverlay, setShouldOpenOverlay] = useState(false);
  const [activeMember, setActiveMember] = useState(null);
  const [activeMemberAbsences, setActiveMemberAbsences] = useState([]);
  const [shouldCleanURLParams, setShouldCleanURLParams] = useState(false);
  const location = useLocation();

  const resetState = useCallback(() => {
    setShouldOpenOverlay(false);
    setShouldCleanURLParams(true);
    setActiveMember(null);
  }, []);

  const downloadICal = useCallback(async () => {
    const absencesToExport = await absencesService.listWithMembers(absences);
    iCalendarService.link({
      absences: absencesToExport,
      filename: 'download.ics'
    });
  }, [absences]);

  const onActiveStartDateChange = useCallback((value) => {
    setSelectedView(value.view);
    setActiveStartDate(value.activeStartDate);
  }, []);

  const onClickDay = useCallback((value) => {
    setSelectedView('day');
    setActiveStartDate(value);
  }, []);

  const onOverlayClose = useCallback(() => {
    resetState();
  }, [resetState]);

  useEffect(() => {
    const responceAbsences = absencesService.list();
    setAbsences(responceAbsences);
  }, []);

  useEffect(() => {
    let absencesToShow;
    switch (selectedView) {
      case 'month':
        absencesToShow = absencesService.getByMonth(absences, activeStartDate);
        break;
      case 'year':
        absencesToShow = absencesService.getByYear(absences, activeStartDate);
        break;
      case 'day':
        absencesToShow = absencesService.getByDate(absences, activeStartDate);
        break;
      default:
        absencesToShow = absencesService.list();
    }
    const absencesWithMembers = absencesService.listWithMembers(absencesToShow);
    setAbsencesToShow(absencesWithMembers);
  }, [absences, selectedView, activeStartDate]);

  useEffect(() => {
    const searchArray = window.location.search.split('=');
    if (searchArray) {
      const userId = searchArray[1] * 1;
      const member = membersService.get(userId);
      if (member) {
        setActiveMember(member);
        setShouldOpenOverlay(true);
        setShouldCleanURLParams(false);
      } else {
        resetState();
      }
    }
  }, [location.search, resetState]);

  useEffect(() => {
    if (activeMember) {
      const memberAbsences = absencesService.getByUser(absences, activeMember.userId);
      setActiveMemberAbsences(memberAbsences);
    }
  }, [activeMember, absences]);

  return (
    <div className="Absences">
      <div onClick={ downloadICal }>
        Download iCal file
      </div>
      <div className="Absences-Container">
        <Calendar
          onClickDay={ onClickDay }
          onActiveStartDateChange={ onActiveStartDateChange }
          value={ new Date(2017, 1, 1, 0, 0) }
        />
        <div className="Absences-List">
          {
            absencesToShow.length > 0
              && absencesToShow.map((absence, key) =>
                <Link className="Absence"
                  key={ `absence-${absence.id}` }
                  to={{
                    pathname: '/',
                    search: `?userId=${absence.userId}`
                  }}
                >
                  <Member
                    id={ key }
                    userId={ absence.userId }
                    name={ absence.user[0].name }
                  />
                  <AbsenceInfo
                    key={ `absenceInfo-${key}` }
                    type={ absence.type }
                    startDate={ absence.startDate }
                    endDate={ absence.endDate }
                    view={ selectedView }
                  />
                </Link>
              )
          }
          {
            absencesToShow.length === 0
            && <p>No absences this {selectedView}</p>
          }
        </div>
      </div>
      {
        shouldOpenOverlay
        && <Overlay
          onClose={ onOverlayClose }
        >
          <Member
            id={ 0 }
            userId={ activeMember.userId }
            name={ activeMember.name }
          />
          <ul>
            {
              activeMemberAbsences.length > 0
              && activeMemberAbsences.map((absence, key) =>
                <AbsenceInfo
                  key={ `absenceInfo-${key}` }
                  type={ absence.type }
                  startDate={ absence.startDate }
                  endDate={ absence.endDate }
                  view={ 'month' }
                />
              )
            }
          </ul>
        </Overlay>
      }
      {
        shouldCleanURLParams
        && <Redirect to="/" />
      }
      <SVGSpriteSheet/>
    </div>
  );
};

export default Absences;