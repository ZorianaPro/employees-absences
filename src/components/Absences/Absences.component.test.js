import React from 'react';
import Absences from './';
import { MemoryRouter as Router } from 'react-router-dom';
import { shallow, mount } from 'enzyme';

describe('Absences', () => {
  let absencesComponent;
  beforeEach(() => {
    absencesComponent = shallow(
      <Router>
        <Absences/>
      </Router>
    );
  });

  it('renders without crashing', () => {
    expect(shallow.bind(
      shallow,
      <Router>
        <Absences />
      </Router>
    )).not.toThrow();
  });

  it('renders as expected', () => {
    expect(absencesComponent.find('.Absences'))
      .toMatchSnapshot();
  });

  describe('absencesToShow', () =>  {
    let absencesComponent;

    beforeEach(() => {
      absencesComponent = mount(
        <Router>
          <Absences/>
        </Router>
      );
    });

    it('renders Absences-List', () => {
      expect(absencesComponent.exists('.Absences-List'))
        .toBe(true);
    });
  });

  // describe('on click on absence', () =>  {
  //   let absencesComponent;
  //
  //   beforeEach(() => {
  //     absencesComponent = mount(
  //       <Router>
  //         <Absences/>
  //       </Router>
  //     );
  //     act(() => {
  //       absencesComponent
  //         .find('.Absence')
  //         .at(0)
  //         .simulate('click')
  //         .update();
  //     })
  //   });
  //
  //   it('renders overlay', () => {
  //     console.log(absencesComponent.find('.Absence a').first().debug())
  //     expect(absencesComponent.exists('.Overlay'))
  //       .toBe(true);
  //   });
  // });
});