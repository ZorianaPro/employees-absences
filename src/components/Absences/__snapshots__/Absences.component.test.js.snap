// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Absences renders as expected 1`] = `null`;
